### ISSUE TITLE
Cypher support for authorization API (part 2)

### ISSUE THREAD

**Issue by adelefelicia** [indykite/neo4j-graph-tool-core#184]
**Created by: @pepe-zamarripa Apr 08 2022 at 15:44 CET**
Last updated: Jul 19 2022 at 14:19 CET
ClickUp ID: IK-2662

Given
IsAuthorized API is called with input parameters:
dt id
list of actions
list of resources refs (ids, type, etc...) 
When
An authorization request is required

Then
output:
for each resource, list of allowed actions

AUTHORISATION API
Static policies (Whenever a new CEO is named that same Permission is assigned to the new guy)
Dynamic policies (management and delegation mechanisms, resource changes, permission context changes)

Extremely-grained Resource link:
poor target set flexibility (high maintenance)
rules always translate to specific nodes.   
(NOTE: when shares/companies are removed, action link is removed. when they are edited, action link stays. The only problem comes when creating new resources. At that point no action link would be present and would need to be created.)

[use the first pic just as reference for the points above. For example permissions are empty in the actual data model with this option]

APOC Subgraph:
flexible resource definition (less maintenance)
rigid rules definition
if labels or relationship types are modified, need to modify the permission specs.
each action could in theory have a separated Permission node (if the actions are not on the same set of resources, or even if those are more prone to change over time).

[in this case we see how the permission link is just one, connecting to a start node.The permission node has the parameters that apoc.subgraph will use to traverse all the nodes indicated from that starting point(s)]

Limitation consideration points:
APOC Subgraph is not flexible enough; actually, we might not find any single suitable procedure, cypher, nor algorithm capable of both generalising and be precise and robust the way we want.
Community Detection, Weak/Strong Connected Components, Subgraph entities (Cliques,etc..), Centrality and Similarity metrics etc.. all give a specific in-perspective hint and resorting to a federation of them would be a better solution.

Final data model consideration points:
We will not consider any Persona or Role atm, just DigitalTwin nodes defining users.
Not enforcing Roles, a DT might have different actions allowed for a same set of resources, in different contexts (Developer and Shares Owner for a Company). Because we are not specifying the context (the Role itself) he's verifying his/her Permissions with, the subgraph procedure will return all his/her allowed actions. We do not want this ambiguity in the future.
About dynamic policies, we will just consider the resource size change (same labels, same relationships, different number of nodes). The "_OWNS" relationship (permission ownership) is not going to be used now, but designed.
The resource reference is a property on the Node and uniquely identifies the resource (node). we need index on that property, at this moment it is called externalId
We model all permission related relationships with a starting "_" (ex: "_ENTITLED"), to be distinguished from the customer-defined ones.
Actions (ex:"_CREATE") are modelled as relationships between the entitled dt and the resource (not nodes nor node attributes)
dt-permission ("_ALLOWED", "_ASSIGNED") --> CHANGED TO "_ENTITLED": still not confident about having the single relationship in the context of a potential additional security check on the double relations and their attributes similarity or exact coincidence.
the Permission parameters are the apoc.subgraph procedure's ones.

Miro frame with data model and query exploration (still to edit and redraw, but contains additional notes). The mixed data model is like:

Of this last model, we consider two versions:
A. It's the one depicted right above, with subgraph params inside the Permission node. If the actions involved different sets of resources, we would have to create one Permission node for each action.
B. We maintain the subgraph params inside the Action relationship, allowing the detachment of the resource definition from the Permission and instead integrate it as an action property(ies).
Additionally, we could argue that we might want or need to have some of the params in the Permission and some in the Action, and their individual position may be dependent on the customer resource set structure

Known Issues:
so there would be a subgraph call for each action relationship defining subgraph params? is this more convenient anyway? ➝ I'd say yes to both
resource not filtered enough ➝ is apoc subgraph better as a coarse filter then? And then refine with other filters ➝ I'd keep it simple for the iteration but consider it for the next 
maybe we are missing something: depth/terminator nodes might not be enough in some cases, and too strict in others. how blacklisting rel types can improve the path's end? ➝ I think blacklisting rel types could be employed easily with the customer because we already have the whitelist. Instead for example, I would not add the filter on the relationship type direction, it creates another logic I'm not sure we want to handle.
in some way we should be able to always have a clear definition of the resource set borders. if not of the resource set itself. what if for example we can start from the right starting nodes instead of one, such that we both are more precise in matching all of them (and not one more), while still being able to be flexible for newcomers/deletions? so that the only operation needed if labels and reltypes stay the same, is to move the secondary starting nodes to include the new resource subgraph.

Additional Observations/Thoughts:
action rels give adv in both detaching actions from permission node in data model and indexing actions 
(permissions all of dt, not of specific "role" requested)➝. think this is not a problem now since the query output would be then diffed with the actually allowed actions, thus recognising the outliers in there
match the starting nodes first and if those are allowed continue with apoc subgraph?
what about propagating information inside the resource graph to comprehend its structure?
what about community logical operators (union, intersection, etc...) after more subgraph calls on the resource graph. For example, use 3 starting nodes being always sure that the intersection of the 3 resulting subgraphs is what you want.
extra mock nodes that define some "wall" or "parameter" to the resource set? But it has to be both flexible and robust..so the borders have to be sensed and moved accordingly..
subgraph constraints: if we know from the customer some specific structural properties of the resource graph that we know won't change if not manually from user's input, we might abstract this metadata about the subgraph structure and not allowing disruptive changes to the structure?
actually, can we find the resource borders by running a combination of simple graph algorithms? For example we might think our border candidates are divided into: peripheral (they're also at the border of the whole connected subgraph component) or central. Measuring and dividing the (in+out) degrees of the component and then the resource subgraph between them should give us an estimation of each node's nature (central or peripheral). I assume the central would be the toughest ones 

FXC Data be like
From Marius B:
Norwegian national identification number (nnin) has following format: DDMMYYXXXXX
The Norwegian national identity number is a unique identifying number that you are assigned for life. It consists of 11 digits, of which the first six digits indicate the person's date of birth.
there will be a property on the digital twin with type nnin and that value.

From Andreas H (FXC):
 Open Policy Agent (OPA) (FXC Platform)
        FXC must set up OPA itself in its own environment.
        The following type of queries are defined via OPA at FXC and run against Indykite Knowledge Graph (This is FXC's graph at Indykite which takes care of relationships and authorization. Also known as "the graph at Indykite" and "FXC Knowledge Graph". Sorry if I switch by name in the summary. Hereinafter known as the "graph" in this summary)
            "Does Person X, have rights to change shareholding Y, in Company Z.?"
            "Can Person X perform action Y on service Z"
            etc

    Delegation of rights
        When the Chairman of the Board delegates rights and roles to others, this is typically something that is represented in the "graph". Typically via "arrows" that go from one object to another object. When the graph has been updated via ingest engine, via API updates etc then OPA should be used to ask if "person X has the right to Y"
            Ex: "Chairman X, delegates writing rights to" investor presentation "to Person Z (ex. Secretary)

Principle:
    FXC must have a "master identification key for persons internally with them (ie in postgresDB)
    FXC must have "represented the most central roles in a company, those that form the basis for a rights model internally with them (ie PostgresDB). Roles are typically chairman of the board and general manager.
    All data points that FXC stores must have metadata stored with them (can be in PostgresDB)
        When was the data last updated?
        "Level of insurance". This is an internal FXC model that says something about how much we trust or value data. Ex. on a scale of 1-4 we can say that data from BRREG is 4, proff is 2, trapeze is 3.
        What is the source of the information? (ex. Organ number from Brreg, Address from Proff.no etc., ownership from Trapets)
        Use; This is to be able to later make queries and choose "the best data source" about quality is important. Especially useful as you can often see that several sources have the same information. Ex. General manager is stored in BRREG, General manager is stored in Proff, General manager is stored with another third party we trust. Then it is important to be able to choose either the source with the highest quality or the most up-to-date.

Identifiers such as NIN and Org Nr that may have specific syntax per country must be taken into account and "generic" variable names should be created that are used. Do not use "social security number" as a variable, rather use "identification number". Based on country, the format of this may be different.

FXC Platform (incl. PostgresDB)
FXC creates a unique identifier for all a). persons or b) organizations that own a share (but may never have logged in), are on the platform or in any other way linked to either companies, shares, or services on the FXC platform. Identifier should be GUID or UUID, Not serial number.
Metadata related to a share is stored in PostgresDB
Metadata related to a company is stored in PostgresDB
Metadata about a person must be considered whether to be stored in PostgresDB or whether it should be stored as an attribute in the person object on the graph. (This is something that can potentially also be downloaded from Indykite.me)
FXC must consider whether a unique identifier should be created or whether we should go "directly to BRØK" and use ID from there. This point needs to be discussed.

FXC Knowledge Graph (hereafter "graph"):
The personal object is created in the "graph".
NIN (in Norway Social security number) is added as an attribute in the personal object on the graph.
Internal Identifier of the person object located in FXC postgres is added as an attribute in the person object on the graph.
The company object is created in the "graph"
The internal identifier of the company object located in the FXC postgres is added as an attribute in the company object on the graph.
The stock object is created in the "graph".
The internal identifier of the share object located in the FXC postgres (possibly FRACTION) is added as an attribute in the company object on the graph. This point needs to be discussed.

From Marius S:
As stated in the text, this is the basic data needed on the knowledge graph for doing authorisation decisions. It's very simple, and the data structures may be a lot more complex, but I think the amount of data types will be something like this.
The naming of the types will change for sure, but I think the meaning will remain.
And the model will not show how the authorisation decisions will look like, or the authorisation API will be like, but I guess you will have an idea.
I have made a diagram that shows how FXC integrates with IndyKite. For this diagram, the focus is on the functionality to control access to users of FXC and how IndyKite can significantly simplify how FXC retrieves information related to relationships between entities.
Access control can generally be divided into two areas. Access is derived from information that comes from a credible source (BRREG for example) and access that is delegated from others who have this authority.
The diagram does not show authentication or authorization, we will return to this later.
Some explanations for the diagram. I have suggested some field names, which you certainly have other names of your own today, but here is my list of terms:

"FXC_ID" - internally generated key that is unique to all users of the platform. A UUID will be a good key.
"FXC_COM_ID" - internally generated key for companies, which, like that for users, is linked to national identifiers (Org. No. In Germany is certainly something different than in Norway).
"FXC_SHARE_ID" - "Grouping" of shares we visited yesterday. Exactly what it is called we find out, and for us it does not matter, just that we know we are talking about the same thing. As the model is today, graph will not provide an answer to how large an ownership fraction a person / company has, only who the owners are. To actually find out the number of shares, the backend to FXC must be asked.
"Sample data" shows what the FXC sends to the IndyKite endpoint. There are rules in Ingest (defined by FXC, but with IndyKite's assistance) that convert this from a "flat" data structure to graph. The format of this will be sent to IndyKite's API and which technology will be used we will agree on later. We will support many different types here, but due. time frames, we want to agree on one. We have the following API technologies in the backlog; REST / JSON, Google pubsub, SQL, CSV, Web hook (JSON) and there are pros and cons with several of them.
"Metadata" arises as a result of rules on "Ingest", and the rules are decided by FXC with assistance from IndyKite. It is not necessary for FXC to have this in its own database. Metadata has not been drawn for all objects, to simplify the diagram.
"Digital Twins" (digital representation of physical identity) in "Identity Graph" (Ann and John) are created based on login with BankID, and all "Properties" associated with them come from BankID, with the exception of "FXC_ID" which enriches Digital Twin based on Ingest (again, based on rule).

The diagram shows that Digital Twin has a reference object in the Knowledge Graph, which we do for compliance reasons, among other things, but this is a complexity that is completely hidden from FXC and you can assume that the node that is colored beige is identical to the one that is brown by integration with the platform.
"Knowlegde Graph API" is used by FXC's backend to perform read / write operations against graph. In this case, there is an interactive session from the FXC frontend where John gives Ann a delegated access "COMPANY_ADMIN". The related "Consent" is signed, and thus verifiable if John denies having given Ann this access.
In the "Knowledge Graph" it is not a goal to duplicate data from other data sources more than necessary. It is the structure between objects and metadata that gives the great value here, and with the help of the identifiers (FXC_COM_ID for example) you can quickly find the depth information in FXC's own backend.
This shows a simple use case, and as you can see, it is a company that has ownership interests in another. These could have been far more advanced structures, but in our graph it will be easy to see that John has a finger in the pie in this structure. This can also be achieved in a SQL database, but then it becomes a static model that must be modeled in advance, and which will be very time-consuming to change if other use cases and structures come in. By e.g. introduction of a new "type" of an object, it is enough to just add this to the set of rules on Ingest, ie no change of code on the platform of IndyKite.
I hope this gives an understanding of how the platform of IndyKite will fit in with FXC.

APOC Subgraph Cypher (unfinished, not used atm, also documented in Miro frame linked above)
params
:params 
{
    resources:['ext_01','ext_02','ext_03','ext_04', 'ext_21', 'ext_31', 'ext_41', 'ext_d11', 'ext_d21'],   
    actions:['_READ']
}
cypher version B (NOT READY)
MATCH (dt:DigitalTwin{id:'001'})-[:_ENTITLED*]->(pe) //add * to search entitled path (role/groups/etc present)
WITH pe
UNWIND  $ actions AS action
MATCH (pe)-[r]->(res)
WHERE TYPE(r)=action
CALL apoc.path.subgraphAll(
 res,
 {
  minLevel: 0,         
  maxLevel: 4,
  relationshipFilter: r.relationshipFilter,
  labelFilter: r.labelFilter
  //bfs:true,
  //filterStartNode:true
 } )
yield nodes, relationships
WITH nodes,relationships,r
UNWIND nodes as res
UNWIND  $ resources AS resource
//WHERE TYPE(r)=action OR res.externalId=resource //This "OR" returns all perms to DT, not the specific one
WITH DISTINCT res.externalId AS resource_id,collect(DISTINCT TYPE(r)) AS actions
WITH collect(apoc.map.fromValues([resource_id, actions])) AS maps
WITH apoc.map.mergeList(maps) AS list
UNWIND list as l
RETURN l

//version A would start one apoc subgraph call from each permission node
//MATCH (dt:DigitalTwin{id:'001'})-[:_ENTITLED]->(pe:Permission)
//WITH pe
//CALL apoc.path.subgraphAll(pe...
with output in this format:
  {
    "res": {
"ext_41": [
        "_READ"
      ],
"ext_01": [
        "_READ"
      ],
"ext_31": [
        "_READ"
      ],
"ext_21": [
        "_READ",
        "_DELETE",
        "_UPDATE",
        "_CREATE"
      ],
"ext_65": [
        "_READ"
      ],
"ext_04": [
        "_READ"
      ],
"ext_05": [
        "_READ"
      ],
"ext_02": [
        "_READ"
      ],
"ext_03": [
        "_READ"
      ],
"ext_d11": [
        "_READ"
      ],
"ext_d21": [
        "_READ"
      ]
    }
  }

APOC Subgraph FG Conclusion:
The investigation done will be continued when having to integrate OPA policies into the graph. The data model used right now will be the Extremely Fine Grained one.
Data Model: started having permission nodes with list of resource.action as one of its properties; linked the permission to the resources and detached both actions and resources from the permission. Started thinking about subgraphs (resources/permission_target and permission related).
Cypher: use (at the moment) cyphers for navigating the semi-known permission subgraph and APOC.subgraphAll procedure to match the almost-unknown resource subgraph. 
_ENTITLED path and some other things may give us a logic-agnostic or flexible graph for how the permissions are assigned between dts, roles/groups. Eventually, also catching nested structures (like Zanzibar or so) of both roles/groups/... and of permissions. Still on the _ENTITLED path, right now I'm obviously just doing a simple pattern match, and then solve the subgraph call for resources. I think that with complexity arising on the permission side, another subgraph call could be run having very strict relationship types available, whatever the node. TOEDIT
Configuration: labelFilter, relationshipFilter, maxLevel
Next: Zanzibar, OPA

ZANZIBAR

Allows clients to define arbitrary relations between users and objects, such as owner, editor, commenter, and viewer. 
In our case this is implemented by rendering these arbitrary relations nodes (could represent roles, groups, personas, etc...). Let's call them User-Object Interfaces (UOIs) for sake of brevity.  

It includes set-algebraic operators such as intersection and union for specifying potentially complex access control policies in terms of those user-object relations. For example, an application can specify that users granted editing rights on a document are also allowed to comment on the document, but not all commenters are given editing rights. 
In our case this is implemented by considering the _ENTITLED path as a sequence of potentially nested UOIs and Permissions, with any combination of those in the sequence possible. We could think of a way to specifically add configurability in terms of operations on sets, but I think we could go beyond that logic. 

At runtime, allows clients to create, modify, and evaluate access control lists (ACLs) through a remote procedure call (RPC) interface. A simple ACL takes the form of “user U has relation R to object O”. More complex ACLs take the form of “set of users S has relation R to object O”, where S is itself specified in terms of another object-relation pair. ACLs can thus refer to other ACLs, for example to specify that the set of users who can comment on a video consists of the users who have been granted viewing rights on that specific video along with those with viewing permissions on the video channel. 
For us, ACLs are how we model data in the lower layer, not at the client level. In such a way, we would already cover both the nested triples (extending it) and the grouping of users (through an UOI). We're also into resource grouping of course.  

Authorization checks take the form of “does user U have relation R to object O?” and are evaluated by a collection of distributed servers. When a check request arrives to Zanzibar, the work to evaluate the check may fan out to multiple servers, for example when a group contains both individual members and other groups. Each of those servers may in turn contact other servers, for example to recursively traverse a hierarchy of group memberships. The ACL data does not lend itself to geographic partitioning because authorization checks for any object can come from anywhere in the world. Therefore, Zanzibar replicates all ACL data in tens of geographically distributed data centers and distributes load across thousands of servers around the world. 
Cannot really say anything about this at the moment.   

It evaluates deeply or widely nested group membership with Leopard, a specialised index for efficient computation of set operations with snapshot consistency. We are working with graphs, so our indexes would be graph-oriented. Starting from the individual nodes and relationship index, we might figure out a way to represent a subgraph index by abstracting the subgraph to be embedded in a node with its structural properties. The (Permission)-[Action]-> couple could be (creatively) thought as a first step on experimenting embedding a subgraph structure (resource subgraph) into a higher-level node. 

OPA OVERVIEW 
TOEDIT

APOC Triggers:
As a final alternative way of achieving an automated assignment of action links to new resources (if desired for that specific permission), I have looked deeper into APOC Triggers. This model has been called TFG (Triggered and Fine-Grained).
The test below clarifies that also these trigger automation is not natively meant in Neo4j to be other than a global graph change automated event triggering. What we are doing here with the triggers instead is to use a one trigger for each action (in a single permission). These triggers depend on global events like "node added to the graph", so we would have for example 100 permissions that fire a different trigger with same event precondition (node added), but then just the ones satisfying more precise conditions (node with labels x and y, connected to node with label x or y, through a relationship of type z) will actually conclude the trigger-related action/cypher (for example create a new permission relationship to the last added node/resource).
We can easily see what the overhead will be, so this is not any viable solution in the mid-run.

TEST SCENARIO
We build a small graph (pictured below), with two DTs.
One of the two, that we're going to use, has two Permissions: one for being CEO and one for being Owner of Shares.

The permissions are fine_grained (one action link from Permission to each resource involved)
And we can ask about the actions on resources like the DT has with:
MATCH (dt:DigitalTwin{id:'001'})-[:_ENTITLED*]->(pe:Permission)
WITH pe
UNWIND  $ actions AS action
UNWIND  $ resources AS resource
MATCH (pe)-[r]->(res)
WHERE TYPE(r)=action OR res.externalId=resource
WITH DISTINCT res.externalId AS resource_id,collect(DISTINCT TYPE(r)) AS actions
WITH collect(apoc.map.fromValues([resource_id, actions])) AS maps
RETURN apoc.map.mergeList(maps) AS res
The results will be then diffed to just return the allowed ones.
Test 1
New resource nodes are added to the graph: four nodes connected to nodes in the _READ path.

Desired Behavior:
each new node directly connected to _READ path node will receive an incoming _READ action link (2)
the indirectly connected will not have any change (1)
How:
We add a permission-action trigger that:
for any new node n (with a label and rel filter) connected to any other already existing node m (with same filter)
if m is also having an inbound relationship named like the action
//and with permission id
then create same relationship going from m to n
The trigger will be like:
CALL apoc.trigger.add('P01|_READ',"UNWIND  $ createdNodes AS n MATCH (n)-[rel]-(m) WHERE (n:Shares OR n:Company) AND TYPE(rel) IN ['OWNS','OF'] AND (m:Shares OR m:Company) WITH n,m MATCH (m)<-[:_READ]-() MERGE (m)-[:_READ]->(n)", {phase:'before'})

//MERGE (m)-[:_READ]->(n) is actually a remnant from the "cascading" action link, the right clause here would be instead MERGE (pe)-[:_READ]->(n) [where pe is the permission, so we have to get it from the permission id (see below) 
//WHERE rel.permId='P01' to be added after last match.

To be complete, the test would need also a way to retrace which Permission(s) this node belongs to. When considering several  Permission on similar set of resources, we would need to: 1) get back the path upwards up to the Permission(s) nodes; 2) propagate a list of Permission ids through the action links to the new action linked to the new node created.
Next:
As a stable solution, we would need to consider way more than APOC Triggers. Actually, Neo4j might not have a native solution for our specific case, thus we would need to code some User Defined Procedure in Java to be installed and then called from within the DB. In general a discussion should be lifted up since the problem is not trivial.

Conclusions:
For this iteration we might use the Extremely Grained Action Link Data Model, with optional roles/groups between dt and permission:
:params {dt_id:'001', resources:['ext_01','ext_02','ext_03','ext_04', 'ext_21', 'ext_31', 'ext_41', 'ext_d11', 'ext_d21'], actions:['_READ']}

MATCH (dt:DigitalTwin{id: $ dt_id})-[:_ENTITLED*]->(pe:Permission)
WITH pe
UNWIND  $ actions AS action
UNWIND  $ resources AS resource
MATCH (pe)-[r]->(res)
WHERE TYPE(r)=action OR res.externalId=resource
WITH DISTINCT res.externalId AS resource_id,collect(DISTINCT TYPE(r)) AS actions
WITH collect(apoc.map.fromValues([resource_id, actions])) AS maps
RETURN apoc.map.mergeList(maps) AS res
to test it properly, we would need a bigger graph than the one I was testing onto. Right now, the small test graph is created with:
//[IK-2662] - create TFG 
CREATE (c1:Company{name:'W',externalId:'ext_01'})
CREATE (c2:Company{name:'Y',externalId:'ext_02'})
CREATE (c3:Company{name:'3W',externalId:'ext_03'})
CREATE (c4:Company{name:'VW',externalId:'ext_04'})
CREATE (c2)-[:OWNS]->(sh21:Shares{qty:5,externalId:'ext_21'})-[:OF]->(c1)
CREATE (c3)-[:OWNS]->(sh31:Shares{qty:33,externalId:'ext_31'})-[:OF]->(c1)
CREATE (c4)-[:OWNS]->(sh41:Shares{qty:50,externalId:'ext_41'})-[:OF]->(c1)
CREATE (c5:Company{name:'X',externalId:'ext_05'})-[:OWNS]->(sh52:Shares{qty:25,externalId:'ext_52'})-[:OF]->(c2)
CREATE (c6:Company{name:'Z3',externalId:'ext_06'})-[:OWNS]->(sh65:Shares{qty:10,externalId:'ext_65'})-[:OF]->(c5)
//dts
CREATE (dt1:DigitalTwin{id:'001'})-[:LIVES_IN]->(n:Country{name:'Norway'})
CREATE (dt1)-[:OWNS]->(shd11:Shares{qty:2,externalId:'ext_d11'})-[:OF]->(c1)
CREATE (dt2:DigitalTwin{id:'002'})-[:OWNS]->(shd21:Shares{qty:10,externalId:'ext_d21'})-[:OF]->(c1)
//permissions
CREATE (dt2)-[:_OWNS]->(pe1:Permission{permId:'P01'})
CREATE (dt1)-[:_ENTITLED]->(:Owner)-[:_ENTITLED]->(pe1)
CREATE (dt1)-[:_ENTITLED]->(:CEO)-[:_ENTITLED]->(pe2:Permission{permId:'P02', labelFilter:'Company|Shares', relationshipFilter:'+OWNS|+OF'})
//actions
//pe1
CREATE (pe1)-[:_READ]->(c1)
CREATE (pe1)-[:_READ]->(shd11)
CREATE (pe1)-[:_DELETE]->(shd11)
//pe2
CREATE (pe2)-[:_CREATE{maxLevel:2,labelFilter:'Shares', relationshipFilter:'+OF'}]->(c5)
CREATE (pe2)-[:_READ{maxLevel:4,labelFilter:'Company|Shares', relationshipFilter:'+OWNS|+OF'}]->(c5)
CREATE (pe2)-[:_READ]->(sh65)
CREATE (pe2)-[:_READ]->(c6)
CREATE (pe2)-[:_READ]->(sh52)
CREATE (pe2)-[:_READ]->(c2)

CREATE (pe2)-[:_UPDATE]->(c5)
CREATE (pe2)-[:_UPDATE]->(sh52)
//CREATE (pe2)-[:_DELETE]->(c5)

as already mentioned, this is not a real solution and it implies high maintenance of the permission links to new resource being added to the set already present.

:warning: Waiting on: 
https://github.com/indykite/jarvis-proto/issues/1470 ,

**Comment by adelefelicia** [comment]
**Commented by: @bayegan Apr 28 2022 at 20:34 CET**

@indyboox : would you be more specific on the TBD sections on the requirements.  will be useful for testing. tnx.

**Comment by adelefelicia** [comment]
**Commented by: @indyboox May 22 2022 at 16:33 CET**

@bayegan @yetizer I have tried to document each step I've taken and concluded by documenting the steps to reproduce the small graph and the sample permission query+params. As extensively written about, this solution does not take into consideration resource set change related to a specific Permission. To achieve this, we simply need more time and some more accurate descriptions of the initial use-cases. To be overly generic is impossible without having:

- some real data from the customer, meaning also the features they want to use;
- policy definition examples in such a way that the data model could reflect the rules and abstract them going on.

I remain available during the week to discuss, modify, extend, model some more data or whatever you might think it would be necessary to get the customer onboard without bad surprises.



Also, do I need to set up a db in dev then so that @andskaal could try it out? If yes, with what data (mock or wait for sort of real one from FXC)? In that case I can coordinate with @maprokes when he would be available. 

Thanks!
